content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+For example, we have a second nodePool with one node. The pool has **debug** and **foo** tags assigned. Now see the options for **foo** tag in **kubernetesLabels**. This means that the m1c43 node will boot with these two labels and taint assigned. Everything seems to be simple. Now [let's try](https://asciinema.org/a/40728) this in practice.
+
+### Demo
+
+Go to [examples](https://github.com/kvaps/kubefarm/tree/master/examples) and update our previous deployed chart to Kubefarm. Just use the [generic](https://github.com/kvaps/kubefarm/tree/master/examples/generic) parameters and look at the pods. We can see that we have added a PXE server and one more job. This job essentially goes to the deployed Kubernetes cluster and creates a new token. Now it will run repeatedly every 12 hours to generate a new token, so that the nodes can connect to our cluster.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot05.jpg)
+
+In a [graphical representation](https://kvaps.github.io/images/posts/Kubernetes-in-Kubernetes-%D0%B8-%D1%84%D0%B5%D1%80%D0%BC%D0%B0-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80%D0%BE%D0%B2-%D1%81-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%BE%D0%B9-%D0%BF%D0%BE-PXE/Argo_CD_Applications_kubefarm-network.html), it looks about the same, but now our apiserver started to be exposed outside. 
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/argocd02.jpg)
+
+In the diagram, the IP is highlighted in green, our PXE server can be reached through it. At the moment, Kubernetes does not allow creating a single LoadBalancer service for TCP and UDP protocols by default, so we have to create two different services with the same IP address. One is for TFTP, and the second for HTTP, through which, the system image is downloaded.
+
+But this simple example is not always enough, sometimes you might need to modify the logic at boot. For example, here is a directory [advanced_network](https://github.com/kvaps/kubefarm/tree/master/examples/advanced_network), inside which there is a [values file](https://github.com/kvaps/kubefarm/tree/master/examples/advanced_network) with a simple shell script. We call it network.sh:
+
+* [**network.sh**](https://github.com/kvaps/kubefarm/blob/3bd90666cfe367e38ee16aa5d6f6f39747874a17/examples/advanced_network/values.yaml#L14-L78)
+
+All this script does is take environment variables at boot time, and generates a network configuration based on them. It creates a directory and puts the netplan config inside. For example, a bonding interface is created here. Basically, this script can contain everything you need. It can hold the network configuration or generate the system services, add some hooks or describe any other logic. Anything that can be described in bash or shell languages will works here, and it will be executed at boot time.
```

Comment:
```suggestion
All this script does is take environment variables at boot time, and generates a network configuration based on them. It creates a directory and puts the netplan config inside. For example, a bonding interface is created here. Basically, this script can contain everything you need. It can hold the network configuration or generate the system services, add some hooks or describe any other logic. Anything that can be described in bash or shell languages will work here, and it will be executed at boot time.
```